fetch.bs

```diff
@@ -4197,32 +4222,68 @@ Range Requests</cite>. [[HTTP-RANGE]] However, this is not widely supported by b
 
       <!-- cache hit -->
       <ol>
-       <li><p>If <var>storedResponse</var> requires validation (i.e., it is not fresh), then set the
-       <var>revalidatingFlag</var>.
-
        <li>
-        <p>If the <var>revalidatingFlag</var> is set and <var>httpRequest</var>'s
-        <a for=request>cache mode</a> is neither "<code>force-cache</code>" nor
-        "<code>only-if-cached</code>", then:
+        <p>If <a for=request>cache mode</a> is "<code>default</code>" and <var>storedResponse</var>
+        is a <a>stale-while-revalidate response</a>, then:
 
         <ol>
-         <li><p>If <var>storedResponse</var>'s <a for=response>header list</a>
-         <a for="header list">contains</a> `<code>ETag</code>`, then <a for="header list">append</a>
-         `<code>If-None-Match</code>` with its value to <var>httpRequest</var>'s
-         <a for=request>header list</a>.
-
-         <li><p>If <var>storedResponse</var>'s <a for=response>header list</a>
-         <a for="header list">contains</a> `<code>Last-Modified</code>`, then
-         <a for="header list">append</a> `<code>If-Modified-Since</code>` with its value to
-         <var>httpRequest</var>'s <a for=request>header list</a>.
+         <li><p>Set <var>response</var> to <var>storedResponse</var>.
+
+         <li><p>Set <var>response</var>'s <a for=response>cache state</a> to "<code>local</code>".
+
+         <li><p>Let <var>revalidateRequest</var> be a <a for=request>clone</a> of
+         <var>request</var>.
+
+         <li><p>Set <var>revalidateRequest</var>'s <a for=request>cache mode</a> set to
+         "<code>no-cache</code>".
+
+         <li><p>Set <var>revalidateRequest</var>'s
+         <a for=request>prevent no-cache cache-control header modification flag</a>.
+
+         <li><p>Set <var>revalidateRequest</var>'s <a for=request>service-workers mode</a> set to
+         "<code>none</code>".
+
+         <li>
+          <p><a>In parallel</a>, perform <a for=main>main fetch</a> using
+          <var>revalidateRequest</var>.
+
+          <p class=note>This fetch is only meant to update the state of the HTTP cache and the
+          response will be unused until another cache access. The stale response will be used as the
+          response to current request. This fetch is issued in the context of a client so if it goes
+          away the request will be terminated.
```

Comment:
How is this enforced? Shouldn't we null check revalidateRequest's client before going here then?

Reply:
I really don't know how to answer this. It is a note so it isn't part of the algorithm just calling out that the context of the revalidation occurs is in the browsing context's scope. Not some other random process.

Reply:
@dtapuska not all fetches have a browsing context. Your note says "This fetch is issued in the context of a client" but that guarantee isn't made in the prose. So, the question is, what do you want to happen to fetches that don't have a client? My gut feeling is that stale-while-revalidate should work there too.

@annevk can you give an example of a fetch without a client? I can never remember

Reply:
OCSP, certain downloads, certain favicon fetches. (Not all of these are well-defined.)

Reply:
And these are precisely things that Chrome's SWR won't work on because they are not fetched in the context of an agent. It will work for agents (workers, windows, service workers). 

@annevk would you feel more comfortable omitting this text entirely?

Reply:
I'd be more comfortable if that was specified, as per my initial comment. 😊

Reply:
Although specifying it would solve the note issue, I'm worried that we're codifying an artefact of Chrome's implementation.

Does it make sense that SWR doesn't work in these cases? Thinking of how I'd explain it to developers, it feels a bit arbitrary.

Reply:
@annevk I've added a not null condition to the if clause.

@jakearchibald That was largely a debate that was before for Chrome where SWR is implemented and having it at the top level (instead of the cache level) had these issues but provides the cancelibility that is desirable when a page terminates all outstanding requests terminate. Developers really shouldn't be relying on caching for functionality. 

For OCSP and downloads it doesn't make sense to do it for because they don't affect the initial display of the page which is what largely this feature targets. Do these requests without a client go through service workers as well?

Reply:
> Developers really shouldn't be relying on caching for functionality.

I guess we'd be telling developers "just because you use stale-while-revalidate doesn't mean the entry in the cache will ever be revalidated".

Reply:
It would still be revalidated but it would just be done after expiry and revalidated synchronously (basically ignoring swr directive) on the next fetch of it. It would be nice to hear from Gecko what they did.  But I downloads nor OCSP likely hit the cache in Chrome anyways so it is probably an even more limited set of fetches that are unknown.

Reply:
I'm also interested in what Gecko did, but given that contextless fetches are rare, I'm happy that the cache isn't updated for these resources (now it's in the spec)

Reply:
@mayhemer does the no client stuff look ok to you? ie. some fetches have no client so we avoid SWR in those cases.

Reply:
Are you discussing fetch() behavior or HTTP request/response handling in general?

The initial implementation in Gecko is very simple, few unimplemented bits will be fixed as followups.  We respect SWR for any request (generally, not just for fetch() initiated requests), in any context, and even w/o a context.  There is no explicit coalescing of multiple revalidateRequests.  The revalidateRequest is not cancelled when the client (source page) is cancelled its load or otherwise goes away.

I have a followup [bug filed](https://bugzilla.mozilla.org/show_bug.cgi?id=1542322) for the cancellation bit.  The exact functionality is to be written down yet.  Cancellation in a context makes sense.  No cancellation for context-less (usually internally initiated system) requests makes sense too.

So, to answer: if there is a context, cancel.  If there is no context, don't cancel.

I personally don't see a reason why we should not respect SWR for some selected requests.  The SWR spec comes from an HTTP RFC, not fetch() spec, and doesn't way anything about context.  SWR can be generally beneficial.

Reply:
@annevk Is this PR change only affecting the fetch() spec?  If so, then the Chrome implementation seems (after discussion on IRC with @dtapuska) reasonable and Gecko will likely end up doing the same/close thing.  

Hence, specifying the non-null client condition to respect SWR for fetch() initiated requests would make sense.

It seems to me pointless to discuss downloads, OCSP and other types of browser initiated internal requests.  Those can potentially benefit from SWR, independently from fetch() and don't need cancellation.

Reply:
@annevk Are we ok to proceed with merging this PR then?

Reply:
We're not, because the fetch algorithm covers all requests, not just those from `fetch()`, right? You can't ever have a `fetch()`-initiated fetch with a null client (except perhaps from browser-privileged code, but that's not specified).

Reply:
@annevk  Can you please advise what changes need to be made then?

Reply:
I'm not sure as I don't understand the constraints well enough. But I think I do understand things to the point where the above exchange ends up being contradictory, as I tried to point out. Was it unclear that the fetch algorithm governs all fetches, not just those made by `fetch()`?

Reply:
@dtapuska it was my understanding that Chrome's implementation of SWR covers all fetches with a client (eg, an iframe's src, an `img`, a script's src etc etc). Is this true?

Reply:
Yes Chrome's implementation covers all fetches with a client. Gecko's implementation does not have that requirement so it will be activated in more scenarios, albeit as @mayhemer points out some scenarios are limited because Gecko uses "no-store" policy for some of those fetches.

Chrome's implementation was done this way so outstanding SWR would be cancelled when a browsing context is destroyed. The Chrome networking team did not want SWR requests on going after a user has terminated the tab.

Reply:
> Gecko's implementation does not have that requirement

Don't requirements come from the spec? I think that's where this thread is getting confused.

Reply:
I talked with Honza out-of-band and it sounds like Firefox will address the mismatches with the specification in an upcoming version.